atomicapp/nulecule/config.py

```diff
@@ -0,0 +1,123 @@
+        self._data = defaultdict(defaultdict)
+        self._data['defaults'] = defaultdict(defaultdict)
+        self._data['answers'] = defaultdict(defaultdict)
+        self._data['cli'] = defaultdict(defaultdict)
+        self._data['runtime'] = defaultdict(defaultdict)
+
+        for scope, data in DEFAULT_ANSWERS.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='defaults')
+        self.set('provider', DEFAULT_PROVIDER, scope=GLOBAL_CONF, source='defaults')
+
+        for scope, data in answers.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='answers')
+
+        for key, value in cli.items():
+            self.set(key, value, scope=GLOBAL_CONF, source='cli')
+
+    def get(self, key, scope=GLOBAL_CONF):
+        """
+        Get the value of a key in a scope. This takes care of resolving
+        the value by going through the PRIORITY order of the various
+        sources of data.
+
+        Args:
+            key (str): Key
+            scope (str): Scope from which to fetch the value for the key
+
+        Returns:
+            Value for the key.
+        """
+        value = None
+        for source in self.PRIORITY:
+            value = self._data[source][scope].get(key)
+            if value:
+                break
+        return value
+
+    def set(self, key, value, scope=GLOBAL_CONF, source='defaults'):
+        """
+        Set the value for a key within a scope along with specifying the
+        source of the value.
+
+        Args:
+            key (str): Key
+            value: Value
+            scope (str): Scope in which to store the value
+            source (str): Source of the value
+        """
+        self._data[source][scope][key] = value
+
+    def context(self, scope=GLOBAL_CONF):
+        """
+        Get context data for the scope of Nulecule graph item. This context
+        data is used to render the artifacts of Nulecule graph item.
+
+        Args:
+            scope (str): Scope (or namespace) for the Nulecule graph item.
+        Returns:
+            A dictionary
+        """
+        result = {}
+        for source in reversed(self.PRIORITY):
+            data = self._data[source]
+            result.update(copy.deepcopy(data.get(GLOBAL_CONF) or {}))
+            result.update(copy.deepcopy(data.get(scope) or {}))
+        return result
+
+    def runtime_answers(self):
+        """
+        Get runtime answers.
+
+        Returns:
+            A defaultdict containing runtime answers data.
+        """
+        answers = defaultdict(dict)
+
+        for source in reversed(self.PRIORITY):
+            for scope, data in (self._data.get(source) or {}).items():
+                answers[scope].update(copy.deepcopy(data))
+
+        # Remove empty sections for answers
+        for key, value in answers.items():
+            if value is None:
+                answers.pop(key, None)
```

Comment:
why `answers.pop(key, None)` vs `answers.pop(key)`. we already know `key` is in the dict right? so there is no risk of `KeyError` being raised